tsdb/engine/tsm1/engine.go

```diff
@@ -972,45 +1022,34 @@ func (e *Engine) compactTSMFull() {
 						files []string
 						err   error
 					)
-					if optimize {
-						files, err = e.Compactor.CompactFast(group)
-						if err == errCompactionsDisabled || err == errCompactionInProgress {
-							e.logger.Printf("aborted %s compaction group (%d). %v",
-								logDesc, groupNum, err)
-
-							if err == errCompactionInProgress {
-								time.Sleep(time.Second)
-							}
-							return
-						} else if err != nil {
-							e.logger.Printf("error compacting TSM files: %v", err)
-							atomic.AddInt64(&e.stats.TSMOptimizeCompactionErrors, 1)
 
-							time.Sleep(time.Second)
-							return
-						}
-					} else {
-						files, err = e.Compactor.CompactFull(group)
-						if err == errCompactionsDisabled || err == errCompactionInProgress {
-							e.logger.Printf("aborted %s compaction group (%d). %v",
-								logDesc, groupNum, err)
-
-							if err == errCompactionInProgress {
-								time.Sleep(time.Second)
-							}
-							return
-						} else if err != nil {
-							e.logger.Printf("error compacting TSM files: %v", err)
-							atomic.AddInt64(&e.stats.TSMFullCompactionErrors, 1)
+					// Count the compaction as active only while the compaction is actually running.
+					func() {
```

Comment:
The side-by-side Github diff is unbearable, but I think the code's maintainability is improved with this PR. It [used to be two copy-pasted chunks of 16 lines with two lines different](https://github.com/influxdata/influxdb/blob/b1d074c01927d241da8fca7ed1a7bd48dfcd95a0/tsdb/engine/tsm1/engine.go#L975-L1026), and it used to [increment the wrong error stat in the optimize case](https://github.com/influxdata/influxdb/blob/b1d074c01927d241da8fca7ed1a7bd48dfcd95a0/tsdb/engine/tsm1/engine.go#L1013), but now [it's less indented with the logic written once instead of twice](https://github.com/influxdata/influxdb/blob/be5a5dee56d4670142cb8f9af9a86fcc75681f20/tsdb/engine/tsm1/engine.go#L1034-L1055).

I don't follow what you mean by taking advantage of nil interfaces here. Golang docs seem to consider a nil interface a variable with an interface type, whose value is nil. I don't think that's what you meant - could you be more specific?